### PR TITLE
revset: add `first_ancestors()` function

### LINKED ISSUE #6606: FR: Add an equivalent to `--first-parent`
**Is your feature request related to a problem? Please describe.**
A project I work on uses (mostly) merge commits to merge branches into `main`, where the shape of the commits in the individual branches is largely up to the developer working on that branch.  When I create a new release, I want to get a list of merge commits and other commits that live directly on `main`.

Currently, in a collocated repo, I can do this with `git log <prev tag>..main --first-parent`.  This has the semantics: "When finding commits to include, follow only the first parent commit upon seeing a merge commit."  (from the [git docs](https://git-scm.com/docs/git-log#Documentation/git-log.txt---first-parent)).

As I understand it, there is currently no equivalent in Jujutsu.

**Describe the solution you'd like**
I suggest a function like `first_parents(srcs, domain)`, which returns all commits in `domain` that can be reached by a commit in `srcs`, following first-parent connections only.

For example, given this history:
```
o F
|
o E
|\
| o D
| |
| o C
| |
o | B
|/
o A
|
o root()
```

then:

* `first_parents(F, A::)` would return `F | E | B | A`
* `first_parents(F, A..)` would return `F | E | B`
* `first_parents(D, A::)` would return `D | C | A` (because even though C is A's second child, A is C's first parent)
* `first_parents(F | D, A::)` would return `F | E | D | C | B | A`  (all commits are either on the first-commit branch or the branch that ends in D)
* `first_parents(E, A::)` would return `E | B | A` (even though `A::` includes F, F is not a parent of E)

**Describe alternatives you've considered**
* A function like `first_parent(X)` that returns only the first parent of `X`.  This might also be useful, but I don't think it's sufficient to get the same results as `--first-parent`, which explicitly *follows* those first-parent relationships recursively.
* This could be an extension of `reachable(srcs, domain)`, i.e. `reachable(srcs, domain, 1)` would be roughly the equivalent of the above.  However, `reachable` currently traverses all parent _and child_ edges, which may be a bit unintuitive when trying to figure out which commits are reachable via first-parent relationships only.  And `reachable(..., 1)` looks weird: what does the `1` refer to here?
* To make this slightly more generic, we could instead have a function like `nth_parents(srcs, domain[, N])`, which would follow the `N`th parent rather than the first parent.  The behaviour is difficult to specify for commits with a single parent though, and I suspect the only two interesting cases are "has a first-parent relationship" and "does not have a first-parent relationship", which are covered by `first_parents(...)` and the usual revset operations.
* This could be a CLI flag like in Git, but that feels like the wrong approach given the revset language exists.

### PR DESCRIPTION
**Pull request by scott2000**
Fixes #6606. Not sure if this is the best approach, but I wanted to try it out.

# Checklist

If applicable:

- [X] I have updated `CHANGELOG.md`
- [X] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [X] I have added/updated tests to cover my changes

### CODE REVIEW

File: lib/src/default_index/rev_walk.rs [hunk 1]
```diff
@@ -247,6 +248,17 @@ pub(super) trait RevWalkIndex {
     type AdjacentPositions: IntoIterator<Item = Self::Position>;
 
     fn adjacent_positions(&self, pos: Self::Position) -> Self::AdjacentPositions;
+
+    fn adjacent_positions_in_range(
```
Comment (scott2000) [inline lib/src/default_index/rev_walk.rs]: If we enforce that `RevsetExpression::Range` uses the same parents range for both the roots and the heads, then we could maybe use a custom `RevWalkIndex` implementation which overrides `adjacent_positions` to only return the parents in a certain range instead.

But I think `first_parents(x) & y..` would be more common than `first_parents(x) ~ first_parents(y)`, so I thought it might be better to leave the unwanted roots using the full set of parents so that this could be optimized to a `RevsetExpression::Range`.

Comment (yuja) [reply lib/src/default_index/rev_walk.rs]: It might be good to add separate `heads_parents_range` and `roots_parents_range`, but I agree we shouldn't use the same range for both heads and roots. The goal of `RevsetExpression::Range` is to omit unwanted ancestors efficiently, and the `roots` and `heads` are usually independent expressions.

**Comment by thoughtpolice** [comment]
nit: I wonder if there's a better name than `first_parents`. I know this sounds (a little) ridiculous, but I somewhat detest compound words for "first-party" revsets, and I could also say the same of e.g. `fork_point`. It shouldn't block this but if anyone has any clever ideas, please post them!

**Comment by yuja** [comment]
There are previous discussions in
https://github.com/jj-vcs/jj/issues/4579 and https://github.com/jj-vcs/jj/issues/6606. If anyone has naming suggestion, it's probably better to continue in either of these issues. FWIW, Mercurial calls it a `p1(x)`.

**Comment by emilazy** [comment]
I think this actually implements `first_ancestors()`. The `first_parents()` we discussed is different. Doesn't solve the compound thing of course.

**Comment by yuja** [comment]
> I think this actually implements `first_ancestors()`. The `first_parents()` we discussed is different. Doesn't solve the compound thing of course.

Good catch. `first_parents(heads, domain)` described originally in #6606 is equivalent to `ancestors(heads, follow_parents_range=0..1) & domain`, and this `first_parents(heads)` implements `ancestors(heads, follow_parents_range=0..1)`? I don't know if we'll need [this version](https://github.com/jj-vcs/jj/issues/6606#issuecomment-2901277705) because it can be expressed by combining more primitive expressions. For the same reason, I don't think we need the original `first_parents(heads, domain)`.

BTW, I misunderstood that this `first_parents(heads)` would compute `ancestors(heads, depth=1..2, follow_parents_range=0..1)` (i.e. `c.parents()[0] for c in heads`.)

**Comment by scott2000** [comment]
I updated it to `first_ancestors(heads)` for now, since it seems like people find `first_parents(heads)` to be a confusing name.

**Comment by scott2000** [comment]
I think this should be ready for review now. I left it as `first_ancestors()` since that seems to be the most preferred option currently.

File: lib/src/default_index/rev_walk.rs [hunk 2]
```diff
@@ -248,6 +249,17 @@ pub(super) trait RevWalkIndex {
...
```
Comment (yuja) [inline lib/src/default_index/rev_walk.rs]: nit: can you extract this to free function? I don't think trait method is needed.

BTW, I think it's simpler to use slice than `.skip()` and `.take()`.

File: lib/src/default_index/rev_walk.rs [hunk 3]
```diff
@@ -248,6 +249,17 @@ pub(super) trait RevWalkIndex {
...
+        &self,
+        pos: Self::Position,
+        parents_range: &Range<u32>,
+    ) -> impl IntoIterator<Item = Self::Position> {
+        self.adjacent_positions(pos)
+            .into_iter()
+            .skip(parents_range.start as usize)
+            .take(parents_range.end.saturating_sub(parents_range.start) as usize)
```
Comment (yuja) [inline lib/src/default_index/rev_walk.rs]: nit: maybe better to normalize the range out of the loop.

File: lib/src/default_index/rev_walk.rs [hunk 4]
```diff
@@ -399,6 +423,10 @@ impl<'a> RevWalkBuilder<'a> {
         self,
         root_positions: HashSet<GlobalCommitPosition>,
     ) -> RevWalkDescendants<'a> {
+        assert_eq!(
+            self.wanted_parents_range, PARENTS_RANGE_FULL,
+            "cannot use parents range with descendants"
+        );
```
Comment (yuja) [inline lib/src/default_index/rev_walk.rs]: nit: I think the output would be correct in that descendants reachable by following the filtered parents range are returned, but I agree it's useless.

File: lib/src/revset.rs [hunk 5]
```diff
@@ -2213,9 +2252,11 @@ fn fold_generation<St: ExpressionState>(
                 RevsetExpression::Ancestors {
                     heads,
                     generation: generation2,
-                } => Some(Rc::new(RevsetExpression::Ancestors {
+                    parents_range: parents2,
+                } if parents2 == parents1 => Some(Rc::new(RevsetExpression::Ancestors {
```
Comment (yuja) [inline lib/src/revset.rs:2300]: Can you update optimization test to include edge cases like this?

**Review by yuja — APPROVED** [review]
LGTM, thanks.

File: lib/src/default_index/rev_walk.rs (hunk 3 above)
Comment (scott2000) [reply lib/src/default_index/rev_walk.rs]: After switching to a slice, I don't think this is applicable anymore.

File: lib/src/default_index/rev_walk.rs (hunk 4 above)
Comment (scott2000) [reply lib/src/default_index/rev_walk.rs]: Yeah, that's true. I removed the assertion since it doesn't seem like an actual error, just something that isn't useful.

File: lib/src/default_index/rev_walk.rs [hunk 6]
```diff
@@ -245,7 +247,7 @@ impl<I, W: RevWalk<I>> FusedIterator for RevWalkOwnedIndexIter<I, W> {}
 
 pub(super) trait RevWalkIndex {
     type Position: Copy + Ord;
-    type AdjacentPositions: IntoIterator<Item = Self::Position>;
+    type AdjacentPositions: IntoIterator<Item = Self::Position> + Deref<Target = [Self::Position]>;
```
Comment (scott2000) [inline lib/src/default_index/rev_walk.rs]: @yuja do you think it's fine to add this `Deref` bound? It makes it so we can use slice operations without having to do `positions.into_iter().collect()` unnecessarily.

Comment (yuja) [reply lib/src/default_index/rev_walk.rs]: I didn't notice the return type was abstract. As a trait bound, `AsRef<[Self::Position]>` should be better.

File: lib/src/default_index/rev_walk.rs [hunk 7]
```diff
@@ -643,6 +668,16 @@ impl RevWalk<CompositeIndex> for RevWalkDescendantsImpl {
     }
 }
 
+/// Filter a slice by a range, clamping the range to the length of the slice.
+fn filter_slice_by_range<'a, T: Copy>(
+    slice: &'a [T],
+    range: &Range<u32>,
+) -> impl Iterator<Item = T> + use<'a, T> {
```
Comment (yuja) [inline lib/src/default_index/rev_walk.rs]: nit: simply return `&'a [T]`?

File: lib/src/default_index/rev_walk.rs [hunk 8]
```diff
@@ -643,6 +668,16 @@ impl RevWalk<CompositeIndex> for RevWalkDescendantsImpl {
...
+    let start = (range.start as usize).min(slice.len());
+    let end = (range.end as usize).min(slice.len());
+    slice[start..end].iter().copied()
```
Comment (yuja) [inline lib/src/default_index/rev_walk.rs]: nit: we'll still need to handle `start > end` somewhere, maybe in `RevWalkBuilder`.
